flocker/volume/filesystems/zfs.py

```diff
@@ -241,39 +299,47 @@ def reader(self, remote_snapshots=None):
         # I'm just using UUIDs, and hopefully requirements will become
         # clearer as we iterate.
         snapshot = b"%s@%s" % (self.name, uuid4())
-        check_call([b"zfs", b"snapshot", snapshot])
+        libzfs_core.lzc_snapshot([snapshot])
 
         # Determine whether there is a shared snapshot which can be used as the
         # basis for an incremental send.
         local_snapshots = list(
             Snapshot(name=name) for name in
-            _parse_snapshots(
-                check_output([b"zfs"] + _list_snapshots_command(self)),
-                self
-            ))
+            _parse_snapshots(_do_list_snapshots(self), self)
+        )
 
         if remote_snapshots is None:
             remote_snapshots = []
 
         latest_common_snapshot = _latest_common_snapshot(
             remote_snapshots, local_snapshots)
-
-        if latest_common_snapshot is None:
-            identifier = [snapshot]
-        else:
-            identifier = [
-                b"-i",
-                u"{}@{}".format(
-                    self.name, latest_common_snapshot.name).encode("ascii"),
-                snapshot,
-            ]
-
-        process = Popen([b"zfs", b"send"] + identifier, stdout=PIPE)
+        latest_common_name = None
+        if latest_common_snapshot is not None:
+            latest_common_name = b"%s@%s" % (self.name,
+                                             latest_common_snapshot.name)
+
+        (rfd, wfd) = os.pipe()
+        out = os.fdopen(rfd)
+        queue = Queue()
+
+        def send_and_close():
+            try:
+                libzfs_core.lzc_send(snapshot, latest_common_name, wfd)
+            except Exception as e:
+                message = ZFS_ERROR(zfs_command="lzc_send " + snapshot,
+                                    output=str(e), status=e.errno)
+                message.write(self.logger)
```

Comment:
> Actually only ZFSError is expected here (Exception is just in case, to not miss the clean-up) and that exception should have enough information. Adding the traceback may be a good idea anyway.

I guess I'm not entirely sure what information it will have.  I know that the exception that used to be handled like this included the output of the `zfs` child process and that was effectively all of the useful information (even though it excluded some Python stack frames having to do with the implementation of running child processes).  Now the exception comes from some cffi-based code and I'm less sure the traceback is irrelevant.  It still could be but I'd like to err on the side of seeing it until we're more confident about that.

Also, mostly separately, on this re-read of the code, I think that this function is being changed so that it no longer actually fails if the zfs send operation has failed.

Prior to the change, if the zfs child process exited with an error code, I think `process.wait()` would have raised an exception and that would have caused this `reader` function to raise an exception.

With the new version of the code, if `lzc_send` raises an exception then it gets logged and then `reader` returns normally - indicating to the caller that the necessary data has been generated.  I think the right thing to do might be to just delete all of this exception handling code and let the exception propagate to the caller of `reader` as it did before?